src/tools/compiletest/src/lib.rs

```diff
@@ -545,46 +543,62 @@ pub fn test_opts(config: &Config) -> test::TestOpts {
     }
 }
 
+/// Read-only context data used during test collection.
+struct TestCollectorCx {
+    config: Arc<Config>,
+    cache: HeadersCache,
+    common_inputs_stamp: Stamp,
+    modified_tests: Vec<PathBuf>,
+}
+
+/// Mutable state used during test collection.
+struct TestCollector {
+    tests: Vec<test::TestDescAndFn>,
+    found_path_stems: HashSet<PathBuf>,
+    poisoned: bool,
```

Comment:
Remark: I want to delete this poison thing and eventually just store a vec of errors/diagnostics, but anyway. Possibly still account for `--fail-fast`, but yeah.

Reply:
Yeah, I had an earlier draft where I tried to get rid of `poisoned` entirely, but in the end I decided to make this PR mostly faithful to the existing behaviour.

At some point in the future, having a list of errors instead of a boolean flag would be great.